relayer/workers/worker.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2021 Snowfork
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package workers
+
+import (
+	"context"
+	"os"
+	"os/signal"
+	"syscall"
+
+	"golang.org/x/sync/errgroup"
+)
+
+type Worker interface {
+	Name() string
+
+	SetUp() error
+	TearDown()
+
+	Start(ctx context.Context, eg *errgroup.Group) error
+}
+
+
+type WorkerPool struct {
+	workers []Worker
+}
+
+func NewWorkerPool(workers []Worker) *WorkerPool {
```

Comment:
I know this still a first-cut WIP, but just adding my thoughts as we flesh this out:

From the looks of it, each worker has persistent state that is re-initialized or mutated by `worker.Start` and `worker.Teardown` between worker restarts. I think it would better if the lifetime of a worker object does not persist between restarts. This would reduce unintended state management bugs, as its a functional approach.

To achieve this, we could add a `pool.RegisterWorker(name, factory)` API, where  `factory` is a function that creates a new `Worker` object on-demand.

Then from the pool's perspective, the worker lifecycle would look something like this:

1. Create worker using factory function
2. Start worker
3. Stop worker if it dies
4. Go back to 1

The factory function could look roughly this:
```
type WorkerFactory func(arg1, arg1) (*Worker, error)
```

Thoughts?

Reply:
yeh sounds nice, though haven't thought about it too deeply

Reply:
A factory sounds good, yes. I'll do that